DUI3-DX/Sdk/Speckle.Autofac/DependencyInjection/SpeckleContainerBuilder.cs

```diff
@@ -0,0 +1,199 @@
+  private readonly IStorageInfo _storageInfo;
+
+  private SpeckleContainerBuilder(IStorageInfo storageInfo, ContainerBuilder? containerBuilder)
+  {
+    _storageInfo = storageInfo;
+    ContainerBuilder = containerBuilder ?? new ContainerBuilder();
+  }
+
+  public static SpeckleContainerBuilder CreateInstance() => new(new StorageInfo(), null);
+
+  private static SpeckleContainerBuilder CreateInstanceForLoading(ContainerBuilder containerBuilder) =>
+    new(new StorageInfo(), containerBuilder);
+
+  // POC: HOW TO GET TYPES loaded, this feels a bit heavy handed and relies on Autofac where we can probably do something different
+  public SpeckleContainerBuilder LoadAutofacModules(Assembly pluginAssembly, IEnumerable<string> dependencyPaths)
+  {
+    // look for assemblies in these paths that offer autofac modules
+    foreach (string path in dependencyPaths)
+    {
+      // POC: naming conventions
+      // find assemblies
+      var assembliesInPath = _storageInfo.GetFilenamesInDirectory(path, "Speckle*.dll");
+      var assemblies = assembliesInPath.Select(LoadAssemblyFile).ToList();
+      if (assemblies.All(x => x != pluginAssembly))
+      {
+        LoadAssembly(pluginAssembly);
+      }
+    }
+
+    return this;
+  }
+
+  private Assembly? LoadAssemblyFile(string file)
+  {
+    try
+    {
+      // inspect the assemblies for Autofac.Module
+      var assembly = Assembly.LoadFrom(file);
+      LoadAssembly(assembly);
+      return assembly;
+    }
+    // POC: catch only certain exceptions
+    catch (Exception ex) when (!ex.IsFatal())
+    {
+      return null;
+    }
+  }
+
+  private void LoadAssembly(Assembly assembly)
+  {
+    var moduleClasses = assembly.GetTypes().Where(x => x.GetInterfaces().Contains(s_moduleType)).ToList();
+
+    // create each module
+    // POC: could look for some attribute here
+    foreach (var moduleClass in moduleClasses)
+    {
+      var module = (ISpeckleModule)Activator.CreateInstance(moduleClass);
+      ContainerBuilder.RegisterModule(new ModuleAdapter(module));
+    }
+  }
+
+  private readonly Lazy<IReadOnlyList<Type>> _types =
+    new(() =>
+    {
+      var types = new List<Type>();
+      foreach (
+        var asm in AppDomain.CurrentDomain
+          .GetAssemblies()
+          .Where(x => x.GetName().Name.StartsWith("Speckle", StringComparison.OrdinalIgnoreCase))
+      )
+      {
+        types.AddRange(asm.GetTypes());
+      }
+      return types;
+    });
+
+  public IReadOnlyList<Type> SpeckleTypes => _types.Value;
+  public ContainerBuilder ContainerBuilder { get; }
+
+  private class ModuleAdapter : Module
+  {
+    private readonly ISpeckleModule _speckleModule;
+
+    public ModuleAdapter(ISpeckleModule speckleModule)
+    {
+      _speckleModule = speckleModule;
+    }
+
+    protected override void Load(ContainerBuilder builder) => _speckleModule.Load(CreateInstanceForLoading(builder));
+  }
+
+  public SpeckleContainerBuilder AddModule(ISpeckleModule module)
```

Comment:
What was the purpose of creating all these `AddModule`, `AddSingleton`, `AddTransient` wrappers?
Why not simply have callers call `ContainerBuilder.RegisterInstance`? 

Seems like we'll have to duplicate a lot of autofac functions and I'm not sure I understand the purpose of this?